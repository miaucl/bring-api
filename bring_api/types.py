"""Bring API types."""
from enum import Enum
from typing import List, TypedDict


class BringList(TypedDict):
    """A list class. Represents a single list."""

    listUuid: str
    name: str
    theme: str


class BringPurchase(TypedDict):
    """A purchase class. Represents a single item."""

    uuid: str
    itemId: str
    specification: str


class BringListItemDetails(TypedDict):
    """An item details class.

    Includes several details of an item in the context of a list.
    Caution: This does not have to be an item that is currently marked as 'to buy'.
    """

    uuid: str
    itemId: str
    listUuid: str
    userIconItemId: str
    userSectionId: str
    assignedTo: str
    imageUrl: str


class BringAuthResponse(TypedDict):
    """An auth response class."""

    uuid: str
    publicUuid: str
    email: str
    name: str
    photoPath: str
    bringListUUID: str
    access_token: str
    refresh_token: str
    token_type: str
    expires_in: int


class BringListResponse(TypedDict):
    """A list response class."""

    lists: List[BringList]


class BringItemsResponse(TypedDict):
    """An items response class."""

    uuid: str
    status: str
    purchase: List[BringPurchase]
    recently: List[BringPurchase]


class BringListItemsDetailsResponse(List[BringListItemDetails]):
    """A response class of a list of item details."""


class BringNotificationType(Enum):
    """Notification type.

    GOING_SHOPPING: "I'm going shopping! - Last chance for adjustments"
    CHANGED_LIST: "List changed - Check it out"
    SHOPPING_DONE: "Shopping done - you can relax"
    URGENT_MESSAGE: "Breaking news - Please get {itemName}!
    """

    GOING_SHOPPING = "GOING_SHOPPING"
    CHANGED_LIST = "CHANGED_LIST"
    SHOPPING_DONE = "SHOPPING_DONE"
    URGENT_MESSAGE = "URGENT_MESSAGE"


class BringNotificationsConfigType(TypedDict):
    """A notification config."""

    arguments: List[str]
    listNotificationType: str
    senderPublicUserUuid: str


class BringUserSettingsEntry(TypedDict):
    """A user settings class. Represents a single user setting."""

    key: str
    value: str


class BringUserListSettingEntry(TypedDict):
    """A user list settings class. Represents a single list setting."""

    listUuid: str
    usersettings: List[BringUserSettingsEntry]


class BringUserSettingsResponse(TypedDict):
    """A user settings response class."""

    usersettings: List[BringUserSettingsEntry]
    userlistsettings: List[BringUserListSettingEntry]


class BringSyncCurrentUserResponse(TypedDict):
    """A sync current user response class."""

    email: str
    name: str
    publicUuid: str
    userLocale: dict[str, str]
    userUuid: str
