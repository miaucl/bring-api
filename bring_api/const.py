"""Constants for bring-api."""
from typing import Final

API_BASE_URL: Final = "https://api.getbring.com/rest/"
DEFAULT_HEADERS: Final = {
    "Authorization": "Bearer",
    "X-BRING-API-KEY": "cof4Nc6D8saplXjE3h3HXqHH8m7VU2i1Gs0g85Sp",
    "X-BRING-CLIENT": "android",
    "X-BRING-APPLICATION": "bring",
    "X-BRING-COUNTRY": "DE",
    "X-BRING-USER-UUID": "",
}

LOCALES_BASE_URL: Final = "https://web.getbring.com/locale/"
BRING_SUPPORTED_LOCALES: Final = [
    "en-AU",
    "de-DE",
    "fr-FR",
    "it-IT",
    "en-CA",
    "nl-NL",
    "nb-NO",
    "pl-PL",
    "pt-BR",
    "ru-RU",
    "sv-SE",
    "de-CH",
    "fr-CH",
    "it-CH",
    "es-ES",
    "tr-TR",
    "en-GB",
    "en-US",
    "hu-HU",
    "de-AT",
]

BRING_DEFAULT_LOCALE: Final = "de-CH"
