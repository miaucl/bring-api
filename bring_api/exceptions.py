"""Bring API exceptions."""


class BringException(Exception):
    """General exception occurred."""


class BringAuthException(BringException):
    """When an authentication error is encountered."""


class BringRequestException(BringException):
    """When the HTTP request fails."""


class BringParseException(BringException):
    """When parsing the response of a request fails."""


class BringEMailInvalidException(BringException):
    """When checkemail returns emailValid false ."""


class BringUserUnknownException(BringException):
    """When checkemail returns userExists false."""


class BringTranslationException(BringException):
    """When translating an article fails."""
