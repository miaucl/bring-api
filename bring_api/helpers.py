"""Helpers for the Bring API."""

import json
from typing import cast


def headers_serialize(headers: dict[str, str]) -> str:
    """Serialize all headers from the login session to a string for persistency.

    Parameters
    ----------
    headers
        The headers in the session to persist

    str
        The serialized data to persist

    """
    return json.dumps(headers)


def headers_deserialize(headers: str) -> dict[str, str]:
    """Deserialize all headers from the login session from a persisted string.

    Parameters
    ----------
    headers
        The persisted session headers as string

    Returns
    -------
    dict[str, str]
        The session cookies

    """
    return cast(dict[str, str], json.loads(headers))
