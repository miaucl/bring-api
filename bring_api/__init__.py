"""Bring API package."""

__VERSION__ = "0.8.1a1"

from .bring import Bring
from .exceptions import (
    BringAuthException,
    BringEMailInvalidException,
    BringParseException,
    BringRequestException,
    BringTranslationException,
    BringUserUnknownException,
)
from .types import (
    BringAuthResponse,
    BringAuthTokenRespone,
    BringItem,
    BringItemOperation,
    BringItemsResponse,
    BringListItemDetails,
    BringListItemsDetailsResponse,
    BringListResponse,
    BringNotificationsConfigType,
    BringNotificationType,
    BringSyncCurrentUserResponse,
    BringUserListSettingEntry,
    BringUserSettingsEntry,
    BringUserSettingsResponse,
)

__all__ = [
    "Bring",
    "BringAuthException",
    "BringAuthResponse",
    "BringAuthTokenRespone",
    "BringEMailInvalidException",
    "BringItem",
    "BringItemOperation",
    "BringItemsResponse",
    "BringListItemDetails",
    "BringListItemsDetailsResponse",
    "BringListResponse",
    "BringNotificationsConfigType",
    "BringNotificationType",
    "BringParseException",
    "BringRequestException",
    "BringSyncCurrentUserResponse",
    "BringTranslationException",
    "BringUserListSettingEntry",
    "BringUserSettingsEntry",
    "BringUserSettingsResponse",
    "BringUserUnknownException",
]
