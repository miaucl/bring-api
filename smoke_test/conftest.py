"""Smoke test for bring-api."""

import logging
import os

import aiohttp
from dotenv import load_dotenv
import pytest

from bring_api.bring import Bring
from bring_api.helpers import headers_deserialize, headers_serialize
from bring_api.types import BringList

load_dotenv()

_LOGGER = logging.getLogger(__name__)


def save_headers(headers: dict[str, str]) -> None:
    """Save the headers locally."""
    with open(".headers", "w", encoding="utf-8") as file:
        file.write(headers_serialize(headers))


def load_headers() -> dict[str, str]:
    """Load the headers locally."""
    # Open and read the file
    with open(".headers", encoding="utf-8") as file:
        return headers_deserialize(file.read() or "{}")


@pytest.fixture(name="headers")
async def cookies_str() -> dict[str, str]:
    """Load the headers."""

    return load_headers()


@pytest.fixture(name="session")
async def aiohttp_client_session():
    """Create  a client session."""
    async with aiohttp.ClientSession() as session:
        yield session


@pytest.fixture(name="bring")
async def bring_api_client(
    session: aiohttp.ClientSession, headers: dict[str, str]
) -> Bring:
    """Create Bring instance."""

    bring = Bring(session, os.environ["EMAIL"], os.environ["PASSWORD"])

    # Restore auth data from saved headers
    bring.headers = headers
    bring.uuid = headers["X-BRING-USER-UUID"]
    bring.public_uuid = headers["X-BRING-PUBLIC-USER-UUID"]
    await bring.reload_user_list_settings()
    await bring.reload_article_translations()

    return bring


@pytest.fixture(name="bring_no_auth")
async def bring_no_auth_api_client(session: aiohttp.ClientSession) -> Bring:
    """Create Bring instance."""

    bring = Bring(session, os.environ["EMAIL"], os.environ["PASSWORD"])

    return bring


@pytest.fixture(name="test_list")
async def test_list(bring: Bring) -> BringList:
    """Get the BringList instance to test with."""
    # Get information about all available shopping list and select the one to test with
    lists = (await bring.load_lists()).lists
    lst = next(lst for lst in lists if lst.name == os.environ["LIST"])
    _LOGGER.info("Selected list: %s (%s)", lst.name, lst.listUuid)
    return lst
