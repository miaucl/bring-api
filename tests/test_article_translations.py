"""Unit tests for bring-api."""

import asyncio
from http import HTTPStatus
from unittest.mock import patch

import aiohttp
from aioresponses import aioresponses
import pytest
from syrupy.assertion import SnapshotAssertion

from bring_api import Bring, BringParseException, BringRequestException, UserLocale
from bring_api.const import BRING_SUPPORTED_LOCALES

from .conftest import UUID

"""Test loading of article translation tables."""


@pytest.mark.usefixtures("mocked")
def test_load_file(
    bring: Bring,
    snapshot: SnapshotAssertion,
) -> None:
    """Test loading json from file."""

    assert bring._Bring__load_article_translations_from_file("de-CH") == snapshot  # type: ignore[attr-defined]


@pytest.mark.usefixtures("mocked")
async def test_load_from_list_article_language(bring: Bring) -> None:
    """Test loading json from listArticleLanguage."""

    with patch.object(
        bring, "user_list_settings", {UUID: {"listArticleLanguage": "de-DE"}}
    ):
        dictionaries = await bring._Bring__load_article_translations()  # type: ignore[attr-defined]

    assert "de-DE" in dictionaries
    assert dictionaries["de-DE"]["Pouletbr체stli"] == "H채hnchenbrust"
    assert len(dictionaries["de-DE"]) == 444


async def test_load_from_user_locale(
    bring: Bring,
) -> None:
    """Test loading json from user_locale."""

    with patch.object(bring, "user_locale", "de-DE"):
        dictionaries = await bring._Bring__load_article_translations()  # type: ignore[attr-defined]

    assert "de-DE" in dictionaries
    assert dictionaries["de-DE"]["Pouletbr체stli"] == "H채hnchenbrust"
    assert len(dictionaries["de-DE"]) == 444


@pytest.mark.parametrize(
    ("test_locale", "expected_locale"),
    [
        ("de-XX", "de-DE"),
        ("en-XX", "en-US"),
        ("es-XX", "es-ES"),
        ("de-DE", "de-DE"),
        ("en-GB", "en-GB"),
    ],
)
async def test_map_user_language_to_locale(
    bring: Bring,
    test_locale: str,
    expected_locale: str,
) -> None:
    """Test mapping invalid user_locale to valid locale."""

    user_locale = UserLocale(language=test_locale[0:2], country=test_locale[3:5])
    locale = bring.map_user_language_to_locale(user_locale)

    assert expected_locale == locale


def test_get_locale_from_list(
    bring: Bring,
) -> None:
    """Test get locale from list."""
    with (
        patch.object(bring, "user_locale", "es-ES"),
        patch.object(
            bring,
            "user_list_settings",
            {UUID: {"listArticleLanguage": "de-DE"}},
        ),
    ):
        locale = bring._Bring__locale(UUID)  # type: ignore[attr-defined]

    assert locale == "de-DE"


def test_get_locale_from_user(
    bring: Bring,
) -> None:
    """Test get locale from user_locale."""
    with (
        patch.object(bring, "user_locale", "es-ES"),
        patch.object(
            bring,
            "user_list_settings",
            {UUID: {"listArticleLanguage": "de-DE"}},
        ),
    ):
        locale = bring._Bring__locale("xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx")  # type: ignore[attr-defined]

    assert locale == "es-ES"


def test_get_locale_from_list_fallback(
    bring: Bring,
) -> None:
    """Test get locale from list and fallback to user_locale."""
    with (
        patch.object(bring, "user_locale", "es-ES"),
        patch.object(bring, "user_list_settings", {UUID: {}}),
    ):
        locale = bring._Bring__locale(UUID)  # type: ignore[attr-defined]

    assert locale == "es-ES"


async def test_load_all_locales(
    bring: Bring,
) -> None:
    """Test loading all locales."""

    user_list_settings = {
        k: {"listArticleLanguage": v} for k, v in enumerate(BRING_SUPPORTED_LOCALES)
    }

    with patch.object(bring, "user_list_settings", user_list_settings):
        dictionaries = await bring._Bring__load_article_translations()  # type: ignore[attr-defined]

    assert len(dictionaries) == 19  # de-CH is skipped


async def test_load_fallback_to_download(
    bring: Bring,
    mocked: aioresponses,
) -> None:
    """Test loading json and fallback to download from web."""
    mocked.get(
        "https://web.getbring.com/locale/articles.de-DE.json",
        payload={"test": "test"},
        status=HTTPStatus.OK,
    )
    with patch("builtins.open", side_effect=OSError):
        await bring.login()

    assert bring._Bring__translations["de-DE"] == {"test": "test"}  # type: ignore[attr-defined]


@pytest.mark.parametrize(
    "exception",
    [
        asyncio.TimeoutError,
        aiohttp.ClientError,
    ],
)
async def test_request_exceptions(
    bring: Bring,
    mocked: aioresponses,
    exception: type[Exception],
) -> None:
    """Test loading json and fallback to download from web."""
    mocked.get(
        "https://web.getbring.com/locale/articles.de-DE.json", exception=exception
    )

    with (
        patch.object(bring, "user_locale", "de-DE"),
        patch("builtins.open", side_effect=OSError),
        pytest.raises(BringRequestException),
    ):
        await bring._Bring__load_article_translations()  # type: ignore[attr-defined]


async def test_parse_exception(
    bring: Bring,
    mocked: aioresponses,
) -> None:
    """Test loading json and fallback to download from web."""
    mocked.get(
        "https://web.getbring.com/locale/articles.de-DE.json",
        status=HTTPStatus.OK,
        body="not json",
        content_type="application/json",
    )

    with (
        patch.object(bring, "user_locale", "de-DE"),
        patch("builtins.open", side_effect=OSError),
        pytest.raises(BringParseException),
    ):
        await bring._Bring__load_article_translations()  # type: ignore[attr-defined]
