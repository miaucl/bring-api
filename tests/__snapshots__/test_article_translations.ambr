# serializer version: 1
# name: test_load_file
  dict({
    'Abfallsäcke': 'Abfallsäcke',
    'Abwaschmittel': 'Abwaschmittel',
    'Adventskalender': 'Adventskalender',
    'Ahornsirup': 'Ahornsirup',
    'Alufolie': 'Alufolie',
    'Ananas': 'Ananas',
    'Anis': 'Anis',
    'Apfelmus': 'Apfelmus',
    'Apfelsaft': 'Apfelsaft',
    'Aprikosen': 'Aprikosen',
    'Aromat': 'Aromat',
    'Artischocken': 'Artischocken',
    'Aubergine': 'Aubergine',
    'Aufbackbrötchen': 'Aufbackbrötchen',
    'Aufschnitt': 'Aufschnitt',
    'Avocado': 'Avocado',
    'BBQ Sauce': 'BBQ Sauce',
    'Babynahrung': 'Babynahrung',
    'Backpapier': 'Backpapier',
    'Backpulver': 'Backpulver',
    'Bacon': 'Bacon',
    'Badreiniger': 'Badreiniger',
    'Baguette': 'Baguette',
    'Ballon': 'Ballon',
    'Balsamico': 'Balsamico',
    'Bananen': 'Bananen',
    'Basilikum': 'Basilikum',
    'Basmatireis': 'Basmatireis',
    'Batterien': 'Batterien',
    'Baumarkt & Garten': 'Baumarkt & Garten',
    'Baumnüsse': 'Baumnüsse',
    'Beeren': 'Beeren',
    'Besteck': 'Besteck',
    'Bier': 'Bier',
    'Binden': 'Binden',
    'Birnen': 'Birnen',
    'Blasenpflaster': 'Blasenpflaster',
    'Blauschimmelkäse': 'Blauschimmelkäse',
    'Blumen': 'Blumen',
    'Blumenerde': 'Blumenerde',
    'Blumenkohl': 'Blumenkohl',
    'Blätterteig': 'Blätterteig',
    'Bodylotion': 'Bodylotion',
    'Bohnen': 'Bohnen',
    'Bouillon': 'Bouillon',
    'Bratensauce': 'Bratensauce',
    'Bratwurst': 'Bratwurst',
    'Bresaola': 'Bresaola',
    'Brezeln': 'Brezeln',
    'Brie': 'Brie',
    'Bring!': 'Bring!',
    'Brokkoli': 'Brokkoli',
    'Brombeeren': 'Brombeeren',
    'Brot': 'Brot',
    'Brot & Gebäck': 'Brot & Gebäck',
    'Brötchen': 'Brötchen',
    'Burrata': 'Burrata',
    'Butter': 'Butter',
    'Buttermilch': 'Buttermilch',
    'Bündnerfleisch': 'Bündnerfleisch',
    'Camembert': 'Camembert',
    'Cervelat': 'Cervelat',
    'Champagner': 'Champagner',
    'Champignons': 'Champignons',
    'Cherrytomaten': 'Cherrytomaten',
    'Chicken Wings': 'Chicken Wings',
    'Chinese To Go': 'Chinesisches Essen',
    'Ciabatta': 'Ciabatta',
    'Cidre': 'Cidre',
    'Cola': 'Cola',
    'Cola Light': 'Cola Light',
    'Conditioner': 'Conditioner',
    'Cordon Bleu': 'Cordon Bleu',
    'Corn Flakes': 'Corn Flakes',
    'Couscous': 'Couscous',
    'Cranberrys': 'Cranberrys',
    'Creme fraiche': 'Creme fraiche',
    'Datteln': 'Datteln',
    'Deo': 'Deo',
    'Dessert': 'Dessert',
    'Dill': 'Dill',
    'Dinkelmehl': 'Dinkelmehl',
    'Dip': 'Dip',
    'Duschmittel': 'Duschmittel',
    'Dörrobst': 'Dörrobst',
    'Dünger': 'Dünger',
    'Eier': 'Eier',
    'Eigene Artikel': 'Eigene Artikel',
    'Eistee': 'Eistee',
    'Eiswürfel': 'Eiswürfel',
    'Energy Drink': 'Energy Drink',
    'Erbsen': 'Erbsen',
    'Erdbeeren': 'Erdbeeren',
    'Erdnüsse': 'Erdnüsse',
    'Essig': 'Essig',
    'Essiggurken': 'Essiggurken',
    'Feigen': 'Feigen',
    'Fenchel': 'Fenchel',
    'Fertig- & Tiefkühlprodukte': 'Fertig- & Tiefkühlprodukte',
    'Festtagsartikel': 'Festtagsartikel',
    'Feta': 'Feta',
    'Feuchttücher': 'Feuchttücher',
    'Fisch': 'Fisch',
    'Fischfutter': 'Fischfutter',
    'Fischstäbli': 'Fischstäbli',
    'Fleisch': 'Fleisch',
    'Fleisch & Fisch': 'Fleisch & Fisch',
    'Fleischwurst': 'Fleischwurst',
    'Fondue': 'Fondue',
    'Frankfurter': 'Wienerli',
    'Frischhaltefolie': 'Frischhaltefolie',
    'Frischkäse': 'Frischkäse',
    'Fruchtsaft': 'Fruchtsaft',
    'Früchte': 'Früchte',
    'Früchte & Gemüse': 'Früchte & Gemüse',
    'Frühlingszwiebeln': 'Frühlingszwiebeln',
    'Fussball-Fest': 'Fussball-Fest',
    'Garnelen': 'Garnelen',
    'Gartenwerkzeug': 'Gartenwerkzeug',
    'Gelee': 'Gelee',
    'Gemüse': 'Gemüse',
    'Gemüse gefroren': 'Gemüse gefroren',
    'Geschenk': 'Geschenk',
    'Geschenkpapier': 'Geschenkpapier',
    'Geschirrglanz': 'Geschirrglanz',
    'Geschirrsalz': 'Geschirrsalz',
    'Geschirrtabs': 'Geschirrtabs',
    'Gesichtscreme': 'Gesichtscreme',
    'Getreideprodukte': 'Getreideprodukte',
    'Getreideriegel': 'Getreideriegel',
    'Getränke': 'Getränke',
    'Getränke & Tabak': 'Getränke',
    'Giesskanne': 'Giesskanne',
    'Gin': 'Gin',
    'Gipfeli': 'Gipfeli',
    'Glacé': 'Glacé',
    'Glasreiniger': 'Glasreiniger',
    'Glühbirne': 'Glühbirne',
    'Glühwein': 'Glühwein',
    'Gnocchi': 'Gnocchi',
    'Gorgonzola': 'Gorgonzola',
    'Gouda': 'Gouda',
    'Granatapfel': 'Granatapfel',
    'Grapefruit': 'Grapefruit',
    'Griess': 'Griess',
    'Grill': 'Grill',
    'Grillartikel': 'Grillartikel',
    'Grillfleisch': 'Grillfleisch',
    'Grissini': 'Grissini',
    'Grüne Bohnen': 'Grüne Bohnen',
    'Grünkohl': 'Grünkohl',
    'Guetzli': 'Guetzli',
    'Gurke': 'Gurke',
    'Haargel': 'Haargel',
    'Haarspray': 'Haarspray',
    'Hackfleisch': 'Hackfleisch',
    'Haferflocken': 'Haferflocken',
    'Hafermilch': 'Hafermilch',
    'Handcreme': 'Handcreme',
    'Haselnüsse': 'Haselnüsse',
    'Haushalt & Gesundheit': 'Haushalt',
    'Haushaltspapier': 'Haushaltspapier',
    'Hefe': 'Hefe',
    'Heidelbeeren': 'Heidelbeeren',
    'Himbeeren': 'Himbeeren',
    'Holzkohle': 'Holzkohle',
    'Honig': 'Honig',
    'Hummus': 'Hummus',
    'Hundefutter': 'Hundefutter',
    'Hundesnack': 'Hundesnack',
    'Hüttenkäse': 'Hüttenkäse',
    'Indian To Go': 'Indisches Essen',
    'Ingwer': 'Ingwer',
    'Insektenschutzmittel': 'Insektenschutzmittel',
    'Italian To Go': 'Italienisches Essen',
    'Joghurt': 'Joghurt',
    'Kaffee': 'Kaffee',
    'Kaffeebohnen': 'Kaffeebohnen',
    'Kaffeekapseln': 'Kaffeekapseln',
    'Kaffeepads': 'Kaffeepads',
    'Kakao': 'Kakao',
    'Kaki': 'Kaki',
    'Kalbfleisch': 'Kalbfleisch',
    'Karotten': 'Karotten',
    'Kartoffeln': 'Kartoffeln',
    'Kartoffelstock': 'Kartoffelstock',
    'Kastanien': 'Marroni',
    'Katzenfutter': 'Katzenfutter',
    'Katzensnack': 'Katzensnack',
    'Katzenstreu': 'Katzenstreu',
    'Kaugummi': 'Kaugummi',
    'Kekse': 'Kekse',
    'Kerzen': 'Kerzen',
    'Ketchup': 'Ketchup',
    'Kichererbsen': 'Kichererbsen',
    'Kidneybohnen': 'Kidneybohnen',
    'Kirschen': 'Kirschen',
    'Kiwi': 'Kiwi',
    'Knoblauch': 'Knoblauch',
    'Knäckebrot': 'Knäckebrot',
    'Kochschinken': 'Kochschinken',
    'Kohl': 'Kohl',
    'Kohlrabi': 'Kohlrabi',
    'Kokosmilch': 'Kokosmilch',
    'Kokosnuss': 'Kokosnuss',
    'Kondome': 'Kondome',
    'Konfitüre': 'Konfitüre',
    'Koriander': 'Koriander',
    'Kosmetiktücher': 'Kosmetiktücher',
    'Kräuter': 'Kräuter',
    'Kräuterbutter': 'Kräuterbutter',
    'Kuchen': 'Kuchen',
    'Kuchenform': 'Kuchenform',
    'Kurkuma': 'Kurkuma',
    'Käse': 'Käse',
    'Kürbis': 'Kürbis',
    'Lachs': 'Lachs',
    'Lamm': 'Lamm',
    'Lasagne': 'Lasagne',
    'Lattich': 'Lattich',
    'Lauch': 'Lauch',
    'Lebensmittelfarbe': 'Lebensmittelfarbe',
    'Leberwurst': 'Leberwurst',
    'Lebkuchen': 'Lebkuchen',
    'Light Limonade': 'Light Limonade',
    'Limette': 'Limette',
    'Limonade': 'Limonade',
    'Linsen': 'Linsen',
    'Linsenmittel': 'Linsenmittel',
    'Lippenpomade': 'Lippenpomade',
    'Lorbeer': 'Lorbeer',
    'Mais': 'Mais',
    'Majoran': 'Majoran',
    'Makeup': 'Makeup',
    'Makeup Entferner': 'Makeup Entferner',
    'Mandarinen': 'Mandarinen',
    'Mandelmilch': 'Mandelmilch',
    'Mandelmus': 'Mandelmus',
    'Mandeln': 'Mandeln',
    'Mango': 'Mango',
    'Margarine': 'Margarine',
    'Mascarpone': 'Mascarpone',
    'Maultaschen': 'Maultaschen',
    'Mayonnaise': 'Mayonnaise',
    'Mehl': 'Mehl',
    'Melone': 'Melone',
    'Milch': 'Milch',
    'Milch & Käse': 'Milch & Käse',
    'Mineralwasser': 'Mineralwasser',
    'Mozzarella': 'Mozzarella',
    'Muffins': 'Muffins',
    'Mundspülung': 'Mundspülung',
    'Muscheln': 'Muscheln',
    'Müesli': 'Müesli',
    'Nagellack': 'Nagellack',
    'Nagellackentferner': 'Nagellackentferner',
    'Nektarine': 'Nektarine',
    'Nelken': 'Nelken',
    'Nougatcreme': 'Nougatcreme',
    'Nudeln': 'Nudeln',
    'Nägel': 'Nägel',
    'Nüsse': 'Nüsse',
    'Oliven': 'Oliven',
    'Olivenöl': 'Olivenöl',
    'Orange': 'Orange',
    'Orangensaft': 'Orangensaft',
    'Orecchiette': 'Orecchiette',
    'Oregano': 'Oregano',
    'Paniermehl': 'Paniermehl',
    'Papaya': 'Papaya',
    'Paprikapulver': 'Paprikapulver',
    'Parmesan': 'Parmesan',
    'Passionsfrucht': 'Passionsfrucht',
    'Pasta': 'Pasta',
    'Pasta Sauce': 'Pasta Sauce',
    'Pelati': 'Pelati',
    'Penne': 'Penne',
    'Peperoncini': 'Peperoncini',
    'Peperoni': 'Peperoni',
    'Pesto': 'Pesto',
    'Petersilie': 'Petersilie',
    'Pfanne': 'Pfanne',
    'Pfeffer': 'Pfeffer',
    'Pfefferkörner': 'Pfefferkörner',
    'Pfefferminze': 'Pfefferminze',
    'Pfirsich': 'Pfirsich',
    'Pflanzen': 'Pflanzen',
    'Pflaster': 'Pflaster',
    'Pflaumen': 'Pflaumen',
    'Pflege & Gesundheit': 'Pflege & Gesundheit',
    'Pilze': 'Pilze',
    'Pinienkerne': 'Pinienkerne',
    'Pinsel': 'Pinsel',
    'Pizza': 'Pizza',
    'Pizzateig': 'Pizzateig',
    'Pommes Chips': 'Pommes Chips',
    'Pommes Frites': 'Pommes Frites',
    'Pop Corn': 'Pop Corn',
    'Poulet': 'Poulet',
    'Pouletbrüstli': 'Pouletbrüstli',
    'Preiselbeer Sauce': 'Preiselbeer Sauce',
    'Propangas': 'Propangas',
    'Prosecco': 'Prosecco',
    'Pudding': 'Pudding',
    'Puddingpulver': 'Puddingpulver',
    'Puderzucker': 'Puderzucker',
    'Putzmittel': 'Putzmittel',
    'Quark': 'Quark',
    'Raclette': 'Raclette',
    'Radieschen': 'Radieschen',
    'Rahm': 'Rahm',
    'Rasenmäher': 'Rasenmäher',
    'Rasierer': 'Rasierer',
    'Rasierklingen': 'Rasierklingen',
    'Rasierschaum': 'Rasierschaum',
    'Ravioli': 'Ravioli',
    'Reibkäse': 'Reibkäse',
    'Reis': 'Reis',
    'Rhabarber': 'Rhabarber',
    'Ricotta': 'Ricotta',
    'Rindfleisch': 'Rindfleisch',
    'Rindsgeschnetzeltes': 'Rindsgeschnetzeltes',
    'Risottoreis': 'Risottoreis',
    'Rohschinken': 'Rohschinken',
    'Rosinen': 'Rosinen',
    'Rosmarin': 'Rosmarin',
    'Rote Bete': 'Randen',
    'Rotwein': 'Rotwein',
    'Rucola': 'Rucola',
    'Rum': 'Rum',
    'Rösti': 'Rösti',
    'Safran': 'Safran',
    'Saft': 'Saft',
    'Salami': 'Salami',
    'Salat': 'Salat',
    'Salatsauce': 'Salatsauce',
    'Salbei': 'Salbei',
    'Salz': 'Salz',
    'Samen': 'Samen',
    'Sardellen': 'Sardellen',
    'Sauerkraut': 'Sauerkraut',
    'Sauerrahm': 'Sauerrahm',
    'Schere': 'Schere',
    'Schinken': 'Schinken',
    'Schmelzkäse': 'Schmelzkäse',
    'Schmerzmittel': 'Schmerzmittel',
    'Schnaps': 'Schnaps',
    'Schnittlauch': 'Schnittlauch',
    'Schnitzel': 'Schnitzel',
    'Schokolade': 'Schokolade',
    'Schrauben': 'Schrauben',
    'Schwamm': 'Schwamm',
    'Schweinefleisch': 'Schweinefleisch',
    'Seife': 'Seife',
    'Sellerie': 'Sellerie',
    'Senf': 'Senf',
    'Servietten': 'Servietten',
    'Shampoo': 'Shampoo',
    'Silberzwielen': 'Silberzwielen',
    'Sirup': 'Sirup',
    'Skyr': 'Skyr',
    'Smoothie': 'Smoothie',
    'Snacks': 'Snacks',
    'Snacks & Süsswaren': 'Snacks & Süsswaren',
    'Socken': 'Socken',
    'Sonnencreme': 'Sonnencreme',
    'Sonnenschirm': 'Sonnenschirm',
    'Soya Milch': 'Soja Drink',
    'Soyasauce': 'Sojasauce',
    'Spaghetti': 'Spaghetti',
    'Spargel': 'Spargel',
    'Speck': 'Speck',
    'Speisestärke': 'Speisestärke',
    'Spinat': 'Spinat',
    'Sportgetränk': 'Sportgetränk',
    'Spätzle': 'Spätzle',
    'Spülmittel': 'Spülmittel',
    'Steak': 'Steak',
    'Stifte': 'Stifte',
    'Streichhölzer': 'Streichhölzer',
    'Strohhalme': 'Strohhalme',
    'Suppe': 'Suppe',
    'Suppengrün': 'Suppengrün',
    'Süssigkeiten': 'Süssigkeiten',
    'Süsskartoffeln': 'Süsskartoffeln',
    'Tampons': 'Tampons',
    'Taschentücher': 'Taschentücher',
    'Tee': 'Tee',
    'Thai To Go': 'Thai Essen',
    'Thon': 'Thon',
    'Thunfisch': 'Thunfisch',
    'Thymian': 'Thymian',
    'Tierbedarf': 'Tierbedarf',
    'Toast': 'Toast',
    'Tofu': 'Tofu',
    'Tomaten': 'Tomaten',
    'Tomatenmark': 'Tomatenmark',
    'Tomatensauce': 'Tomatensauce',
    'Tonic Water': 'Tonic Water',
    'Torte': 'Torte',
    'Tortellini': 'Tortellini',
    'Tortilla Chips': 'Tortilla Chips',
    'Trauben': 'Trauben',
    'Töpfe': 'Töpfe',
    'Vanille': 'Vanille',
    'Vanille Sauce': 'Vanille Sauce',
    'Vanillezucker': 'Vanillezucker',
    'Vielen Dank': 'Vielen Dank',
    'Vitamine': 'Vitamine',
    'Vodka': 'Vodka',
    'Vogelfutter': 'Vogelfutter',
    'Vorgeschlagene Artikel': 'Vorgeschlagene Artikel',
    'WC-Papier': 'WC-Papier',
    'WC-Reiniger': 'WC-Reiniger',
    'WG-Artikel': 'WG-Artikel',
    'Waffeln': 'Waffeln',
    'Waschmittel': 'Waschmittel',
    'Wasser': 'Wasser',
    'Wassermelone': 'Wassermelone',
    'Watterondellen': 'Watterondellen',
    'Wattestäbchen': 'Wattestäbchen',
    'Weichspüler': 'Weichspüler',
    'Weisswein': 'Weisswein',
    'Whisky': 'Whisky',
    'Windeln': 'Windeln',
    'Wurst': 'Wurst',
    'Zahnbürsten': 'Zahnbürsten',
    'Zahnpasta': 'Zahnpasta',
    'Zahnseide': 'Zahnseide',
    'Zahnstocher': 'Zahnstocher',
    'Zigaretten': 'Zigaretten',
    'Zigarren': 'Zigarren',
    'Zimt': 'Zimt',
    'Zitrone': 'Zitrone',
    'Zitronensaft': 'Zitronensaft',
    'Zopf': 'Zopf',
    'Zucchetti': 'Zucchetti',
    'Zucker': 'Zucker',
    'Zuletzt verwendet': 'Zuletzt verwendet',
    'Zutaten & Gewürze': 'Zutaten & Gewürze',
    'Zwetschgen': 'Zwetschgen',
    'Zwiebeln': 'Zwiebeln',
    'getrocknete Tomaten': 'Getrocknete Tomaten',
    'Äpfel': 'Äpfel',
    'Öl': 'Öl',
  })
# ---
